### ISSUE TITLE
Is there any intention of implementing IMPALA?

### ISSUE THREAD

**Issue by tarungog** [astooke/rlpyt#84]
FAIR has an implementation in Pytorch here if it would help -- https://github.com/facebookresearch/torchbeast/

**Comment by codelast** [comment]
rlpyt can only be used in one single machine, from my point of view, I think rlpyt's owner will not make it run on multiple machines.
Comparing to rlpyt, the PolyBeast, which is "somewhat harder to install" according to FAIR, I think rlpyt is really too too too easy to install/setup and more friendly.

**Comment by tarungog** [comment]
A single node implementation of IMPALA is also included, and could be helpful

**Comment by astooke** [comment]
Hi!  This is right that rlpyt as it stands is single-node only, and there's only one set of neural net parameters.  I'd have to go back and look more closely but I think this means the importance sampling in IMPALA would all just go to 1?  This could change if running in asynchronous mode, where there is one optimizer agent and one sampling agent, which can be slightly lagged.  This can probably be handled by PPO.  (note: async policy gradient isn't currently implemented, would need to make PPO build a replay buffer and learn from it like DQN)

Open to other thoughts/reasons, though!

**Comment by tarungog** [comment]
@astooke I think the workflow would be to create aysnchronous sampling for PG, then implement V-trace algorithm and importance sampling. Totally fine if this is not in the roadmap. Just wondering

**Comment by astooke** [comment]
Ok good question!  It's crossed my mind to implement it, but is not yet a priority.  

Would need to find a compelling use case (i.e. the problem is big) in policy gradient where the time spent sampling is roughly equal to the time spent optimizing, so the asynchronous mode could give up to a 2x speedup.  This was the case for the R2D2 re-implementation, where it brought experiment time down to ~140 hours.  It might also take some effort to tune replay hyperparameters to keep good learning performance...something that had been done for us already in the R2D2 work.

**Comment by tarungog** [comment]
Yeah I think Impala is stronger for the multi-node case

**Comment by astooke** [comment]
Right, no immediate plans for IMPALA, closing this for now but please reopen if a new use case comes up :)
